File: core/src/main/java/org/evergreen_ils/views/search/AdvancedSearchActivity.java
/*
 * Copyright (C) 2012 Evergreen Open-ILS
 * @author Daniel-Octavian Rizea
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * or the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software 
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 * 
 */
package org.evergreen_ils.views.search;

import static org.evergreen_ils.ConstKt.KEY_SEARCH_TEXT;
import static org.evergreen_ils.views.search.SearchActivity.RESULT_CODE_SEARCH_BY_KEYWORD;

import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup.LayoutParams;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import net.kenstir.hemlock.R;
import net.kenstir.hemlock.android.App;
import net.kenstir.hemlock.android.ui.ActionBarUtils;
import net.kenstir.hemlock.android.Analytics;

import java.util.ArrayList;
import java.util.StringTokenizer;

public class AdvancedSearchActivity extends AppCompatActivity {

    private final static String TAG = AdvancedSearchActivity.class.getSimpleName();

    private ArrayList<String> searchTerms;
    private ArrayList<String> searchTermTypes;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Analytics.initialize(this);
        if (!App.isStarted()) {
            App.restartApp(this);
            return;
        }

        setContentView(R.layout.advanced_search);
        ActionBarUtils.initActionBarForActivity(this);

        searchTerms = new ArrayList<>();
        searchTermTypes = new ArrayList<>();

        final LinearLayout layout = findViewById(R.id.advanced_search_filters);
        Button addFilter = findViewById(R.id.advanced_search_add_filter_button);
        final Spinner search_class_spinner = findViewById(R.id.advanced_search_class_spinner);
        final Spinner search_contains_spinner = findViewById(R.id.advanced_search_contains_spinner);
        final EditText search_filter_text = findViewById(R.id.advanced_search_text);

        addFilter.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {

                int contains_pos = search_contains_spinner.getSelectedItemPosition();
                String query = "";
                String qtype = search_class_spinner.getSelectedItem().toString().toLowerCase();
                searchTermTypes.add(qtype);
                String filter = search_filter_text.getText().toString();

                switch (contains_pos) {
                case 0:
                    // contains
                    query = qtype + ": " + filter;
                    break;
                case 1:
                    // excludes
                    query = qtype + ":";
                    StringTokenizer str = new StringTokenizer(filter);
                    while (str.hasMoreTokens()) {
                        String token = str.nextToken(" ");
                        query = query + " -" + token;
                    }
                    break;
                case 2:
                    // matches exactly
                    query = qtype + " \"" + filter + "\"";
                    break;
                }
                searchTerms.add(query);
                TextView text = new TextView(AdvancedSearchActivity.this);
                text.setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT));
                text.setText(query);
                layout.addView(text);
            }
        });

        Button cancel = findViewById(R.id.advanced_search_cancel);
        cancel.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
//                Analytics.logEvent("advsearch_cancel");
                finish();
            }
        });

        Button search = findViewById(R.id.advanced_search_button);
        search.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                String types = TextUtils.join("|", searchTermTypes);
                //Analytics.logEvent("advsearch_search", "search_type", types);//TODO
                Intent returnIntent = new Intent();
                returnIntent.putExtra(KEY_SEARCH_TEXT, TextUtils.join(" ", searchTerms));
                setResult(RESULT_CODE_SEARCH_BY_KEYWORD, returnIntent);
                finish();
            }
        });

    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: core/src/main/java/net/kenstir/ui/view/search/AdvancedSearchActivity.kt
/*
 * Copyright (C) 2012 Evergreen Open-ILS
 * @author Daniel-Octavian Rizea
 *
 * Copyright (c) 2025 Kenneth H. Cox
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <https://www.gnu.org/licenses/>.
 */
package net.kenstir.ui.view.search

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.os.bundleOf
import net.kenstir.data.model.SearchClass
import net.kenstir.hemlock.R
import net.kenstir.ui.Lifecycle
import net.kenstir.ui.Key
import net.kenstir.util.Analytics
import java.util.Locale
import java.util.StringTokenizer

class AdvancedSearchActivity: AppCompatActivity() {
    private val searchTerms = mutableListOf<String>()
    private val searchTermTypes = mutableListOf<String>()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (!Lifecycle.isStarted) {
            Lifecycle.restartApp(this)
            return
        }

        Analytics.initialize(this)

        setContentView(R.layout.advanced_search)

        val layout = findViewById<LinearLayout>(R.id.advanced_search_filters)
        val addFilter = findViewById<Button>(R.id.advanced_search_add_filter_button)
        val search_class_spinner = findViewById<Spinner>(R.id.advanced_search_class_spinner)
        val search_contains_spinner = findViewById<Spinner>(R.id.advanced_search_contains_spinner)
        val search_filter_text = findViewById<EditText>(R.id.advanced_search_text)

        initSearchClassSpinner(search_class_spinner)

        addFilter.setOnClickListener {
            val contains_pos = search_contains_spinner.selectedItemPosition
            var query = ""
            val filter = search_filter_text.text.toString().trim()
            if (filter.isEmpty()) {
                return@setOnClickListener
            }
            val qtype = search_class_spinner.selectedItem.toString().lowercase(Locale.getDefault())
            searchTermTypes.add(qtype)

            when (contains_pos) {
                0 -> {
                    // contains
                    query = "$qtype: $filter"
                }
                1 -> {
                    // excludes
                    query = "$qtype:"
                    val str = StringTokenizer(filter)
                    while (str.hasMoreTokens()) {
                        val token = str.nextToken(" ")
                        query = "$query -$token"
                    }
                }
                2 -> {
                    // matches exactly
                    query = "$qtype: \"$filter\""
                }
            }
            searchTerms.add(query)
            val text = TextView(this@AdvancedSearchActivity)
            text.layoutParams =
                ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT)
            text.text = query
            layout.addView(text)
        }

        val cancel = findViewById<Button>(R.id.advanced_search_cancel)
        cancel.setOnClickListener {
            Analytics.logEvent(Analytics.Event.SEARCH_ADV_CANCEL)
            finish()
        }

        val search = findViewById<Button>(R.id.advanced_search_button)
        search.setOnClickListener {
            Analytics.logEvent(Analytics.Event.SEARCH_ADV_SEARCH, bundleOf(
                Analytics.Param.NUM_ITEMS to searchTerms.size,
            ))
            val returnIntent = Intent()
            returnIntent.putExtra(Key.SEARCH_TEXT, TextUtils.join(" ", searchTerms))
            setResult(SearchActivity.RESULT_CODE_SEARCH_BY_KEYWORD, returnIntent)
            finish()
        }
    }

    private fun initSearchClassSpinner(searchClassSpinner: Spinner) {
        val labels = SearchClass.spinnerLabels
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, labels)
        searchClassSpinner.adapter = adapter
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }
}
